sound/soc/sof/sof-audio.c

```diff
@@ -325,6 +327,63 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
 		}
 	}
 
+	/*
+	 * The above loop handles connections between widgets that belong to the DAPM widget list.
+	 * But this is not sufficient to handle special cases such as the one needed for setting
+	 * up the route between the feedback pipeline and the playback pipeline for the smart
+	 * amplifier. So handle those separately while checking that both pipelines are in use.
+	 */
+	list_for_each_entry(sroute, &sdev->route_list, list) {
+		bool src_widget_in_dapm_list, sink_widget_in_dapm_list;
+
+		if (sroute->setup)
+			continue;
+
+		src_widget_in_dapm_list = widget_in_list(list, sroute->src_widget->widget);
+		sink_widget_in_dapm_list = widget_in_list(list, sroute->sink_widget->widget);
+
+		/*
+		 * if both source and sink are in the DAPM list, the route must already have been
+		 * set up above. And if neither are in the DAPM list, the route shouldn't be
+		 * handled now.
+		 */
+		if ((src_widget_in_dapm_list && sink_widget_in_dapm_list) ||
+		    (!src_widget_in_dapm_list && !sink_widget_in_dapm_list))
+			continue;
+
+		/*
+		 * Check the use_count of the widget that is not in the DAPM list to confirm if
+		 * it is in use currently before setting up the route.
+		 */
+		if (src_widget_in_dapm_list) {
+			mutex_lock(&sroute->sink_widget->setup_mutex);
+			if (!sroute->sink_widget->use_count) {
+				mutex_unlock(&sroute->sink_widget->setup_mutex);
+				continue;
+			}
+		} else {
+			mutex_lock(&sroute->src_widget->setup_mutex);
+			if (!sroute->src_widget->use_count) {
+				mutex_unlock(&sroute->src_widget->setup_mutex);
+				continue;
+			}
+		}
+
+		if (tplg_ops && tplg_ops->route_setup) {
+			ret = tplg_ops->route_setup(sdev, sroute);
+			if (!ret)
+				sroute->setup = true;
+		}
+
+		if (src_widget_in_dapm_list)
+			mutex_unlock(&sroute->sink_widget->setup_mutex);
+		else
+			mutex_unlock(&sroute->src_widget->setup_mutex);
+
+		if (ret < 0)
+			return ret;
```

Comment:
question: In the loop above we recursively follow the paths. Here we have a different way of walking through the graph by detecting the possible routes.

This means there's no real order in how these loopback routes are handled. They will be processed in the order they are found in the linked list, without any priority or structure approach based on TBD criteria. Is this ok?

Isn't there a case where setting up a route between two widgets would require us to also do a recursion? My brain is foggy on what a route_setup does and if it has any side effects that would require yet another pass in the graph.

Reply:
I think its OK to set up routes in any order @plbossart. This happens during hw_params and just because there's a route connected doesnt mean that the pipeline on the other end is active. We handle this gracefully with the mixin/mixout. Hopefully the smartamp does the same too.

And Im not sure about the recursion for setting up a route. It is simply just one IPC in most cases.